File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-files-exist.ts
#!/usr/bin/env bun
import type { Finding, HarnessCheckRule, HarnessManifest, RuleContext } from "../harness-check-rule";

/**
 * Require specified files to exist.
 */
export const requireFilesExistRule = (ctx: RuleContext): HarnessCheckRule => ({
  applies(manifest: HarnessManifest): boolean {
    const section = manifest.requireFilesExist;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    if (enabled === false) {
      return false;
    }
    const entry = ctx.readJsonObject((section as Record<string, unknown>).parameters);
    return ctx.readStringArray(entry.paths).length > 0;
  }

  validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
    const section = ctx.readJsonObject(manifest.requireFilesExist);
    const parameters = ctx.readJsonObject(section.parameters);
    const paths = ctx.readStringArray(parameters.paths);
    return paths.flatMap((path) => {
      if (ctx.isSymlink(path) && ctx.allowedRootContractTarget(path) === null) {
        return [
          {
            severity: ctx.severityOf(manifest, "requireFilesExist"),
            category: "requireFilesExist",
            message: `symlink file is not allowed: ${path}`,
          },
        ];
      }
      return ctx.isFile(path)
        ? []
        : [
            {
              severity: ctx.severityOf(manifest, "requireFilesExist"),
              category: "requireFilesExist",
              message: `missing file: ${path}`,
            },
          ];
    });
  }

});

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_files_exist.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require files exist rule."""

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import is_safe_file, severity_for


class RequireFilesExistRule(HarnessCheckRule):
    """Validate requireFilesExist check."""

    category = "requireFilesExist"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> list[Finding]:
        """Validate requireFilesExist check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        paths = params.get("paths", [])
        if not isinstance(paths, list):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        template = messages.get("default", "missing file: {path}")
        return [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                template.format(path=path),
            )
            for path in paths
            if isinstance(path, str) and not is_safe_file(root / path)
        ]
